### PR TITLE
Add Azure Workload Identity support & read groups from claim

### PR DESCRIPTION
**Pull request by JacekLakis-TomTom**
## Description
This pull request enables to use oauth2-proxy with federated credentials offered by Azure Workload Identity. Instead of `client_secret`, every request includes a [certificate credential](https://learn.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#request-an-access-token-with-a-certificate-credential) that is mounted inside the pod with Workload Identity enabled.

In addition, this PR introduces reading groups from id_token claim for V2 tokens instead of Microsoft Graph API, so the app registration does not need the admin consent. They can be retrieved in the same way as in V1 - read `groups` claim. For both cases this claim contains list of UUIDs. That's why I added condition to call Graph API only when groups were not present in id_token.

## Motivation and Context
Currently, oauth2-proxy needs to have hardcoded secret configured, which causes security issues and requires maintenance like secret rotation. This PR closes https://github.com/oauth2-proxy/oauth2-proxy/issues/1979

## How Has This Been Tested?
Custom image has been built and put on AKS cluster with Workload Identity support enabled. Following configuration was provided:
```
        - --http-address=0.0.0.0:4180
        - --metrics-address=0.0.0.0:44180
        - --provider=azure
        - --oidc-issuer-url=https://login.microsoftonline.com/{tenant-id}/v2.0 # tested with V1 as well
        - --email-domain="*"
        - --cookie-domain=".domain.com"
        - --whitelist-domain="*"
        - --footer="Some msg"
        - --skip-jwt-bearer-tokens=true
        - --set-xauthrequest=true
        - --silence-ping-logging=true
        - --reverse-proxy=true
        - --oidc-email-claim=email
        - --insecure-oidc-allow-unverified-email=true
        - --skip-provider-button=true
        - --upstream=static://202
        - --set-authorization-header=false
        - --client-id={client-id}
        - --cookie-secret={cookie-secret}
        - --azure-federated-token-auth-enabled=true # tested with client secrets as well
```
+ Tested login flow.

## Checklist:

Please advise on CHANGELOG entries.

- [ ] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [X] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: docs/docs/configuration/overview.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ An example [oauth2-proxy.cfg](https://github.com/oauth2-proxy/oauth2-proxy/blob/
 | `--auth-logging-format` | string | Template for authentication log lines | see [Logging Configuration](#logging-configuration) |
 | `--authenticated-emails-file` | string | authenticate against emails via file (one per line) | |
 | `--azure-tenant` | string | go to a tenant-specific or common (tenant-independent) endpoint. | `"common"` |
+| `--azure-use-workload-identity` | bool | whether to use Azure AD App federated credential to access Azure from within Kubernetes | false |
```
Comment (JoelSpeed) [inline docs/docs/configuration/overview.md:84]: Why specifically Kube?

File: providers/azure.go [hunk 2]
```diff
@@ -224,14 +227,26 @@ func (p *AzureProvider) prepareRedeem(redirectURL, code, codeVerifier string) (u
...
+		if err != nil {
+			return params, err
+		}
+		params.Add("client_secret", clientSecret)
+	} else {
+		federatedTokenPath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+		federatedToken, err := os.ReadFile(federatedTokenPath)
+		if err != nil {
+			logger.Errorf("error reading federated token file %s: %s", federatedTokenPath, err)
+			return params, err
```
Comment (JoelSpeed) [inline providers/azure.go:242]: Would be better to use `fmt.Errorf` to wrap the error here and add more context

File: providers/azure.go [hunk 3]
```diff
@@ -224,14 +227,26 @@ func (p *AzureProvider) prepareRedeem(redirectURL, code, codeVerifier string) (u
...
+
+	if !p.useWorkloadIdentity {
+		clientSecret, err := p.GetClientSecret()
+		if err != nil {
+			return params, err
+		}
+		params.Add("client_secret", clientSecret)
+	} else {
+		federatedTokenPath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
+		federatedToken, err := os.ReadFile(federatedTokenPath)
```
Comment (JoelSpeed) [inline providers/azure.go:239]: Do we need to read the file every time the app processes the authentication of a user? How often is this expected to change?

File: providers/azure.go [hunk 4]
```diff
@@ -224,14 +227,26 @@ func (p *AzureProvider) prepareRedeem(redirectURL, code, codeVerifier string) (u
...
-		return params, err
+
+	if !p.useWorkloadIdentity {
+		clientSecret, err := p.GetClientSecret()
+		if err != nil {
+			return params, err
+		}
+		params.Add("client_secret", clientSecret)
+	} else {
+		federatedTokenPath := os.Getenv("AZURE_FEDERATED_TOKEN_FILE")
```
Comment (JoelSpeed) [inline providers/azure.go:238]: Maybe check this var isn't empty?

**Review by JoelSpeed — COMMENTED** [review]
Do we need any additional documentation for setting this feature up? Perhaps for links to workload identity and how it needs to be configured?

File: docs/docs/configuration/overview.md (hunk 1 above)
Comment (JacekLakis-TomTom) [reply docs/docs/configuration/overview.md:84]: I would say that's the most common use case, but indeed it's not just kubernetes-specific. Will change it to something more generic.

File: providers/azure.go (hunk 3 above)
Comment (JacekLakis-TomTom) [reply providers/azure.go:239]: Yes, federated credentials are supposed to be rather short-lived and updates in place. They are injected into the pod via Admission control.

**Comment by JacekLakis-TomTom** [comment]
Hey @JoelSpeed , I've added documentation and some other minor changes. I would like to see you review again. Please let me know about any concerns or things to carefuly test.

**Comment by djsly** [comment]
thanks @JacekLakis-TomTom  for this contrib, cannot wait to test it!

**Comment by danjo133** [comment]
I would also very much like to see this support added

**Comment by czhujer** [comment]
Hi @JoelSpeed , can you pls re-open this issue? 😍

**Comment by SebastianBakala-TomTom** [comment]
Can't wait to use it!

**Comment by mblaschke** [comment]
any progress?

File: docs/docs/configuration/auth.md [hunk 5]
```diff
@@ -104,6 +105,34 @@ in the App registration manifest file.
    --oidc-issuer-url=https://login.microsoftonline.com/{tenant-id}/v2.0
 ```
 
+#### Authentication with federated credentials on Kubernetes
```
Comment (kvanzuijlen) [inline docs/docs/configuration/auth.md:108]: I think this would be the prefered method? Should be above the long lived variant.

**Comment by y6vmeq** [comment]
It is great. I also wait this improvments.

File: docs/docs/configuration/auth.md [hunk 6]
```diff
@@ -104,6 +105,34 @@ in the App registration manifest file.
...
+
+And start `oauth2-proxy` with federated credentials enabled:
+```
+    --provider=azure
+    --oidc-issuer-url=https://login.microsoftonline.com/{tenant-id}/v2.0
+    --client-id={client-id}
+    --azure-federated-token-auth-enabled=true
+
+``` 
+Note: Federated credentials can be configured to trust different parties that Azure Kubernetes Service
```
Comment (kvanzuijlen) [inline docs/docs/configuration/auth.md:134]:
```suggestion
Note: Federated credentials can be configured to trust different parties than Azure Kubernetes Service
```

**Comment by mblaschke** [comment]
@JacekLakis-TomTom do you still maintain this PR? I'm interested in this feature to avoid using client secrets.

**Comment by mblaschke** [comment]
@SebastianBakala-TomTom
can you ping @JacekLakis-TomTom? i'm willing to take over this PR if it's not managed anymore

**Comment by jjlakis** [comment]
Hey @mblaschke I am revisiting this right now, hoping to have a new PR ready this year

**Comment by jjlakis** [comment]
@JoelSpeed @djsly @SebastianBakala-TomTom @mblaschke @kvanzuijlen I have revistied this changes and started a new Pull Request, fixing the code climate changes now :point_right: https://github.com/oauth2-proxy/oauth2-proxy/pull/2364
This PR can be closed.

**Comment by tuunit** [comment]
PR will be closed in favour of #2364

**Pull request closed without merging**